src/Analyzer/Passes/QueryAnalysisPass.cpp

```diff
@@ -8212,6 +8212,10 @@ void QueryAnalyzer::resolveQuery(const QueryTreeNodePtr & query_node, Identifier
     {
         auto node = node_with_duplicated_alias;
         auto node_alias = node->getAlias();
+
+        /// Add current alias to non cached set, because in case of cyclic alias identifier should not be substituted from cache.
+        /// See 02896_cyclic_aliases_crash.
+        scope.non_cached_identifier_lookups_during_expression_resolve.insert({Identifier{node_alias}, IdentifierLookupContext::EXPRESSION});
```

Comment:
Resolve caching introduces a lot of problems, it's interesting if it has a big performance impact. Maybe we should try to disable it at all in a separate PR and check perf tests. WDYT?

Reply:
Resolve caching is kind of ok, cause it caches by raw ptr...
Identifier lookup cache is hard to support because it is not aware of the context, and we need to care about cyclic aliases
Alias cache (especially since we store the resolved and unresolved tree in the same map) is scary, and makes our query tree a DAG.